README
MooseX-ConsistentHash version 0.01
==================================

NAME
    MooseX::ConsistentHash - unified consistent hashing interface Version

SYNOPSIS
    some examples

        use MooseX::ConsistentHash;
    
        my $hashing = MooseX::ConsistentHash->new();
        -- or --
        my $hashing = MooseX::ConsistentHash->new(
            class => 'MooseX::ConsistentHash::AlgorithmConsistentHashKetama'
        );
        -- or --
        use Digest::JHash qw();
        my $hashing = MooseX::ConsistentHash->new(
            class => 'MooseX::ConsistentHash::SetConsistentHash', 
            init_hash => {hash_func => \&Digest::JHash::jhash}
        );
        -- or --
        my $hashing = MooseX::ConsistentHash->new(
            class => 'MooseX::ConsistentHash::SetConsistentHash' 
        );
    
        $hashing->add_options(
            {option => 'foo1', weight => 1},
            {option => 'foo2', weight => 2}
        );
        my $some = Some->new();
        $hashing->add_option(option => $some, weight => 10);
    
        my $result_option = $hashing->get_option('some_string');
        # $result_option may be 'foo1' or 'foo2' or $some

DESCRIPTION
    MooseX::ConsistentHash provides a unified consistent hashing API. By
    default, this module can work with
    Algorithm::ConsistentHash::Ketama,Set::ConsistentHash and any of your
    module for consistent hashing.

  METHODS
   new
    It takes two parameters:

    *   class - name class for consistent hashing. The class must support the
        role of MooseX::ConsistentHash::RoleInterface. By default class is
        MooseX::ConsistentHash::AlgorithmConsistentHashKetama

    *   init_hash - optional hashref argument. He will be transferred to a
        constructor consistent hashing class.

   add_option(option => $some, weight => $integer_weight)
    This method added option to the consistent hashing instance through
    MooseX::ConsistentHash::Option object. Return identificator option or
    undef.

   add_options({option => $some, weight => $integer_weight},...)
    Some sugar. Return list identificators options.

   new_weight($identificator, $new_weight)
    change weight for current identificator. retuns true if succeful.

   clear_options
    remove all options and reinint instance consistent hashing. retuns true if
    succeful.

   clear_option($identificator)
    remove one option. retuns true if succeful.

   show_options
    return hashref all options (keys are identificators)

   get_option($key)
    select the option in the set to which that key is mapped and returns
    value.

PROBLEMS
    MooseX::ConsistentHash with
    MooseX::ConsistentHash::AlgorithmConsistentHashKetama or
    MooseX::ConsistentHash::SetConsistentHash slower by 80% than the original
    module Algorithm::ConsistentHash::Ketama or Set::ConsistentHash.

SEE ALSO
    *   MooseX::ConsistentHash::RoleInterface, MooseX::ConsistentHash::Option

    *   MooseX::ConsistentHash::AlgorithmConsistentHashKetama,
        Algorithm::ConsistentHash::Ketama

    *   MooseX::ConsistentHash::SetConsistentHash, Set::ConsistentHash

AUTHOR
    Sivirinov Ivan, <catamoose at yandex.ru>

COPYRIGHT AND LICENSE
    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself, either Perl version 5.10 or, at your
    option, any later version of Perl 5 you may have available.

    And see Algorithm::ConsistentHash::Ketama, Set::ConsistentHash.



Changes
Revision history for Perl extension MooseX::ConsistentHash.

0.01  Sat Sep 25 17:23:57 2010
	- original version; created by h2xs 1.23 with options
		-AX MooseX::ConsistentHash



Makefile.PL
use 5.010001;
use ExtUtils::MakeMaker;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
    NAME              => 'MooseX::ConsistentHash',
    VERSION_FROM      => 'lib/MooseX/ConsistentHash.pm', # finds $VERSION
    PREREQ_PM         => {
                           Digest::MD5 => 0,
                           Moose => 0,
                           Moose::Util::TypeConstraints => 0,
                           Moose::Role => 0,
                           Algorithm::ConsistentHash::Ketama => 0,
                           Set::ConsistentHash => 0
    }, # e.g., Module::Name => 1.1
    ($] >= 5.005 ?     ## Add these new keywords supported since 5.005
      (#ABSTRACT_FROM  => 'lib/MooseX/ConsistentHash.pm', # retrieve abstract from module
       AUTHOR         => 'Sivirinov Ivan <catamoose@yandex.ru>') : ()),
);


lib/MooseX/ConsistentHash/Option.pm
package MooseX::ConsistentHash::Option;

our $VERSION = '0.01';

use Storable qw(freeze);
use Digest::MD5 qw(md5_hex);

use Moose;
use Moose::Util::TypeConstraints;
use namespace::autoclean;

# type describing value for 'hash' attribute
subtype 'md5_hex'
    => as 'Str'
    => where {
        /^[\d|a-z]{32}$/
    }
    => message { "'identificator' value ('$_') of the option is not like a result Digest::MD5::md5_hex\n"};

# option value to choice return
has 'option' => (
    is => 'ro',
    isa => 'Defined',
    required => 1
);

# weight option
has 'weight' => (
    is => 'rw',
    isa => 'Int',
    required => 1
);

# option uniquie identificator (md5 by value thing)
# Probably a good idea to use Data::UUID, but there are doubts in the case of threads or forks
# because the attribute is 'lazy'
# Therefore used a bunch of Digest::MD5 with Storable
has 'identificator' => (
    is => 'ro',
    isa => 'md5_hex',
    lazy => 1,
    default => sub {md5_hex freeze shift}
);

__PACKAGE__->meta->make_immutable;

1;
__END__

=head1 NAME

MooseX::ConsistentHash::Option - is the bucket for MooseX::ConsistentHash::*
I<Version 0.01>

=head1 SYNOPSIS
.
    use MooseX::ConsistentHash::Option;
    my $option = MooseX::ConsistentHash::Option->new(
        value => $some_value,
        weight => $some_integer
    );

=head1 METHODS

In addition to methods 'value' and 'weight' is available:

=head2 identificator

is a 'lazy' method. By default, returns the 'md5_hex' string from the 'freeze' objects value.

=head1 SEE ALSO

L<MooseX::ConsistentHash>

=head1 AUTHOR

Sivirinov Ivan, E<lt>catamoose at yandex.ruE<gt>

=head1 COPYRIGHT AND LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10 or,
at your option, any later version of Perl 5 you may have available.

And see L<Set::ConsistentHash>.

=cut


lib/MooseX/ConsistentHash/RoleInterface.pm
package MooseX::ConsistentHash::RoleInterface;

our $VERSION = '0.01';

use Moose::Role;

# set option
requires 'add_option';

# get choiced identificator
requires 'get_option';


1;
__END__

=head1 NAME

MooseX::ConsistentHash::RoleInterface - abstract interface for MooseX::ConsistentHash::*
I<Version 0.01>

=head1 DESCRIPTION

If you want to add your own MooseX::ConsistentHash::* module, which will be available for use in L<MooseX::ConsistentHash>, then you should use this class as a role.

=head2 Abstaract methods

=head3 add_option($option)

option is a L<MooseX::ConsistentHash::Option> object.
$option-E<gt>identificator and $option-E<gt>weight are all you need.  

=head3 get_option($thing)

'thing' is a string for which your module should return the appropriate option.
You must return $option-E<gt>identificator 

=head1 EXAMPLE

=over

=item 

L<MooseX::ConsistentHash::AlgorithmConsistentHashKetama>

=item 

L<MooseX::ConsistentHash::SetConsistentHash>

=back

=head1 SEE ALSO

=over

=item 

L<MooseX::ConsistentHash>

=item 

L<MooseX::ConsistentHash::Option>

=back

=head1 AUTHOR

Sivirinov Ivan, E<lt>catamoose at yandex.ruE<gt>

=head1 COPYRIGHT AND LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10 or,
at your option, any later version of Perl 5 you may have available.

=cut


lib/MooseX/ConsistentHash/SetConsistentHash.pm
package MooseX::ConsistentHash::SetConsistentHash;

our $VERSION = '0.01';

use Set::ConsistentHash;
use Digest::MD5 qw();

use Moose;
use namespace::autoclean;

with 'MooseX::ConsistentHash::RoleInterface';

has 'hash_func' => (
    is => 'ro',
    isa => 'Any',
    default => sub { return \&Digest::MD5::md5_hex},
    required => 1   
);

has 'consistent_hash' => (
    is => 'ro',
    isa => 'Set::ConsistentHash',
    lazy => 1,
    default => sub {
        my $self = shift;
        my $hash = Set::ConsistentHash->new;
        $hash->set_hash_func($self->hash_func);
        $hash;
    }
);

sub add_option {
    my ($self,$option) = @_;
    $self->consistent_hash->set_target($option->identificator => $option->weight);
    return 1;
};

sub get_option {
    shift->consistent_hash->get_target(@_);
};

__PACKAGE__->meta->make_immutable;

1;
__END__

=head1 NAME

MooseX::ConsistentHash::SetConsistentHash - moose version Set::ConsistentHash for MooseX::ConsistentHash
I<Version 0.01>

=head1 SYNOPSIS

for example

    use Digest::JHash qw();
    use MooseX::ConsistentHash;
    my $hashing = MooseX::ConsistentHash->new(
        class => 'MooseX::ConsistentHash::SetConsistentHash', 
        init_hash => {hash_func => \&Digest::JHash::jhash}
    );
    $hashing->add_options(
        {option => 'foo1', weight => 1},
        {option => 'foo2', weight => 2}
    );
    my $some = Some->new();
    $hashing->add_option(option => $some, weight => 10);
    my $result_option = $hashing->get_option('some_string');
    # $result_option may be 'foo1' or 'foo2' or $some  


=head2 MORE EXAMPLE

another example

    use MooseX::ConsistentHash;
    my $hashing = MooseX::ConsistentHash->new(
        class => 'MooseX::ConsistentHash::SetConsistentHash'
    );
    $hashing->add_options(
        {option => 'foo1', weight => 1},
        {option => 'foo2', weight => 2}
    );
    my $some = Some->new();
    $hashing->add_option(option => $some, weight => 10);
    my $result_option = $hashing->get_option('some_string');
    # $result_option may be 'foo1' or 'foo2' or $some 

By default, MooseX::ConsistentHash::SetConsistentHash used as hash_func Digest::MD5::md5_hex

=head1 PROBLEMS

L<MooseX::ConsistentHash> with L<MooseX::ConsistentHash::SetConsistentHash> slower by 80% than the module L<Set::ConsistentHash>

=head1 SEE ALSO

=over

=item 

L<MooseX::ConsistentHash>

=item 

L<Set::ConsistentHash>

=back

=head1 AUTHOR

Sivirinov Ivan, E<lt>catamoose at yandex.ruE<gt>

=head1 COPYRIGHT AND LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10 or,
at your option, any later version of Perl 5 you may have available.

And see L<Set::ConsistentHash>.

=cut


lib/MooseX/ConsistentHash/AlgorithmConsistentHashKetama.pm
package MooseX::ConsistentHash::AlgorithmConsistentHashKetama;

our $VERSION = '0.01';

use Algorithm::ConsistentHash::Ketama;

use Moose;
use namespace::autoclean;

with 'MooseX::ConsistentHash::RoleInterface';

has 'ketama' => (
    is => 'ro',
    isa => 'Algorithm::ConsistentHash::Ketama',
    lazy => 1,
    default => sub {Algorithm::ConsistentHash::Ketama->new}
);

sub add_option {
    my ($self,$option) = @_;
    $self->ketama->add_bucket($option->identificator, $option->weight);
    return 1;
};

sub get_option {
    shift->ketama->hash(@_);
};

__PACKAGE__->meta->make_immutable;

1;
__END__

=head1 NAME

MooseX::ConsistentHash::AlgorithmConsistentHashKetama - moose version Algorithm::ConsistentHash::Ketama for MooseX::ConsistentHash
I<Version 0.01>

=head1 SYNOPSIS

for example

    use MooseX::ConsistentHash;
    my $hashing = MooseX::ConsistentHash->new(
        class => 'MooseX::ConsistentHash::AlgorithmConsistentHashKetama'
    );
    $hashing->add_options(
        {option => 'foo1', weight => 1},
        {option => 'foo2', weight => 2}
    );
    my $some = Some->new();
    $hashing->add_option(option => $some, weight => 10);
    my $result_option = $hashing->get_option('some_string');
    # $result_option may be 'foo1' or 'foo2' or $some  


=head2 MORE EXAMPLE

another example

    use MooseX::ConsistentHash;
    my $hashing = MooseX::ConsistentHash->new();
    $hashing->add_options(
        {option => 'foo1', weight => 1},
        {option => 'foo2', weight => 2}
    );
    my $some = Some->new();
    $hashing->add_option(option => $some, weight => 10);
    my $result_option = $hashing->get_option('some_string');
    # $result_option may be 'foo1' or 'foo2' or $some 

By default, MooseX::ConsistentHash used as 'class' MooseX::ConsistentHash::AlgorithmConsistentHashKetama.

=head1 PROBLEMS

L<MooseX::ConsistentHash> with L<MooseX::ConsistentHash::AlgorithmConsistentHashKetama> slower by 80% than the module L<Algorithm::ConsistentHash::Ketama>

=head1 SEE ALSO

=over

=item 

L<MooseX::ConsistentHash>

=item 

L<Algorithm::ConsistentHash::Ketama>

=back

=head1 AUTHOR

Sivirinov Ivan, E<lt>catamoose at yandex.ruE<gt>

=head1 COPYRIGHT AND LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10 or,
at your option, any later version of Perl 5 you may have available.

And see L<Algorithm::ConsistentHash::Ketama>.

=cut


lib/MooseX/ConsistentHash.pm
package MooseX::ConsistentHash;

our $VERSION = '0.01';

use Moose;
use Moose::Util::TypeConstraints;
use namespace::autoclean;

use MooseX::ConsistentHash::Option;

# store consistent object after create instance of class
has '_consistent_object' => (
    is => 'rw',
    isa => 'Object',
    lazy => 1,
    default => sub {
        my $self = shift;
        $self->has_hash_option ? $self->class->new($self->hash_option) : $self->class->new();
    }
);

# type describing the class of ConsistentHash with load, if necessary
subtype 'ConsistentClass'
    => as 'ClassName'
    => where {
        $_->does('MooseX::ConsistentHash::RoleInterface');
    }
    => message { "Class '$_' must have role 'MooseX::ConsistentHash::RoleInterface'\n"};

coerce 'ConsistentClass'
    => from 'Str'
    => via {
        Class::MOP::load_class($_); $_;
    };

# the only attribute that should make the constructor, the other attributes are established methods
has 'class' => (
    is => 'ro',
    isa => 'ConsistentClass',
    default => 'MooseX::ConsistentHash::AlgorithmConsistentHashKetama',
    coerce => 1
);

# somr option to init consistent hashing object
has 'hash_option' => (
    is => 'ro',
    isa => 'HashRef',
    predicate => 'has_hash_option'
);

# hold all options
has '_options' => (
    is => 'rw',
    isa => 'HashRef',
    lazy => 1,
    default => sub{ {} }
);

sub add_option {
    my $self = shift;
    my $option = UNIVERSAL::isa($_[0],'MooseX::ConsistentHash::Option') ? $_[0] : MooseX::ConsistentHash::Option->new(@_);
    unless(exists $self->_options->{$option->identificator}) {
        # new option
        return unless $self->_consistent_object->add_option($option);
        $self->_options->{$option->identificator} = $option;
    } elsif ($self->_options->{$option->identificator}->weight ne $option->weight) {
        $self->clear_option($option->identificator);
        return unless $self->_consistent_object->add_option($option);
        $self->_options->{$option->identificator} = $option;
    };
    return $option->identificator;
}

sub add_options {
    my $self = shift;
    return [map {$self->add_option($_)} @_] 
}

sub clear_options {
    my $self = shift;
    $self->_consistent_object($self->has_hash_option ? $self->class->new($self->hash_option) : $self->class->new);
    $self->_options({});
    return 1;
}

sub clear_option {
    my ($self, $identificator) = @_;
    return unless $identificator;
    if (delete $self->_options->{$identificator}) {
        my $options = $self->_options;
        $self->clear_options();
        $self->add_options(values %$options);
        return 1;
    };
    return;
}

sub get_option {
    my $self = shift;
    my $identificator = $self->_consistent_object->get_option(@_);
    return exists $self->_options->{$identificator} ? $self->_options->{$identificator}->option : undef; 
}

sub show_options {
    shift->_options;
}

sub new_weight {
    my ($self, $identificator, $new_weight) = @_;
    return unless ($identificator and int($new_weight) and exists $self->_options->{$identificator});

    $self->_options->{$identificator}->weight($new_weight);
    my $options = $self->_options;
    $self->clear_options();
    $self->add_options(values %$options);    
    
    return 1;
}

__PACKAGE__->meta->make_immutable;
1;
__END__

=head1 NAME

MooseX::ConsistentHash - unified consistent hashing interface
I<Version 0.01>

=head1 SYNOPSIS

some examples

    use MooseX::ConsistentHash;
    
    my $hashing = MooseX::ConsistentHash->new();
    -- or --
    my $hashing = MooseX::ConsistentHash->new(
        class => 'MooseX::ConsistentHash::AlgorithmConsistentHashKetama'
    );
    -- or --
    use Digest::JHash qw();
    my $hashing = MooseX::ConsistentHash->new(
        class => 'MooseX::ConsistentHash::SetConsistentHash', 
        init_hash => {hash_func => \&Digest::JHash::jhash}
    );
    -- or --
    my $hashing = MooseX::ConsistentHash->new(
        class => 'MooseX::ConsistentHash::SetConsistentHash' 
    );
    
    $hashing->add_options(
        {option => 'foo1', weight => 1},
        {option => 'foo2', weight => 2}
    );
    my $some = Some->new();
    $hashing->add_option(option => $some, weight => 10);
    
    my $result_option = $hashing->get_option('some_string');
    # $result_option may be 'foo1' or 'foo2' or $some 

=head1 DESCRIPTION

MooseX::ConsistentHash provides a unified consistent hashing API.
By default, this module can work with L<Algorithm::ConsistentHash::Ketama>,L<Set::ConsistentHash> and any of your module for consistent hashing.

=head2 METHODS

=head3 new

It takes two parameters:

=over

=item 

class - name class for consistent hashing. The class must support the role of L<MooseX::ConsistentHash::RoleInterface>.
By default class is L<MooseX::ConsistentHash::AlgorithmConsistentHashKetama>

=item 

init_hash - optional hashref argument. He will be transferred to a constructor consistent hashing class. 

=back

=head3 add_option(option => $some, weight => $integer_weight)

This method added option to the consistent hashing instance through L<MooseX::ConsistentHash::Option> object.
Return identificator option or undef.

=head3 add_options({option => $some, weight => $integer_weight},...)

Some sugar.
Return list identificators options.

=head3 new_weight($identificator, $new_weight)

change weight for current identificator.
retuns true if succeful.

=head3 clear_options

remove all options and reinint instance consistent hashing.
retuns true if succeful.

=head3 clear_option($identificator)

remove one option.
retuns true if succeful.

=head3 show_options

return hashref all options (keys are identificators)

=head3 get_option($key)

select the option in the set to which that key is mapped and returns value.

=head1 PROBLEMS

L<MooseX::ConsistentHash> with L<MooseX::ConsistentHash::AlgorithmConsistentHashKetama> or L<MooseX::ConsistentHash::SetConsistentHash> slower by 80% than the original module L<Algorithm::ConsistentHash::Ketama> or L<Set::ConsistentHash>.

=head1 SEE ALSO

=over

=item 

L<MooseX::ConsistentHash::RoleInterface>, L<MooseX::ConsistentHash::Option>

=item 

L<MooseX::ConsistentHash::AlgorithmConsistentHashKetama>, L<Algorithm::ConsistentHash::Ketama>

=item 

L<MooseX::ConsistentHash::SetConsistentHash>, L<Set::ConsistentHash>

=back

=head1 AUTHOR

Sivirinov Ivan, E<lt>catamoose at yandex.ruE<gt>

=head1 COPYRIGHT AND LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.10 or,
at your option, any later version of Perl 5 you may have available.

And see L<Algorithm::ConsistentHash::Ketama>, L<Set::ConsistentHash>.

=cut